src/serial.rs

```diff
@@ -0,0 +1,484 @@
+//! Serial mock implementations
+//!
+//! We may expect serial read and write transactions on a mock
+//! Serial device. Note that, in the `embedded_hal` crate, there
+//! are the non-blocking serial traits, and there is a blocking
+//! variant. We may use the same mock for both interfaces.
+//!
+//! ## Usage: non-blocking serial traits
+//!
+//! ```
+//! extern crate embedded_hal;
+//! extern crate embedded_hal_mock;
+//!
+//! // Note that we're using the non-blocking serial traits
+//! use embedded_hal::serial::{Read, Write};
+//! use embedded_hal_mock::serial::{
+//!     Mock as SerialMock,
+//!     Transaction as SerialTransaction,
+//! };
+//!
+//! // Configure expectations
+//! let expectations = [
+//!     SerialTransaction::read(0x0A),
+//!     SerialTransaction::read(0xFF),
+//!     SerialTransaction::write([1, 2]), // (1)
+//!     SerialTransaction::flush(),
+//! ];
+//!
+//! let mut serial = SerialMock::new(&expectations);
+//!
+//! // Expect two reads
+//! let word = serial.read().unwrap();
+//! assert_eq!(word, 0x0A);
+//! let word = serial.read().unwrap();
+//! assert_eq!(word, 0xFF);
+//!
+//! // When designing against the non-blocking serial
+//! // trait, we expect two separate writes. These could be
```

Comment:
could or should?

Reply:
Could is correct. Given the nuances of the implementation, either of the expectations

```
let batch = [
    SerialTransaction::write([1, 2]),
];

let serial = [
    SerialTransaction::write([1]),
    SerialTransaction::write([2]),
];
```
could satisfy the following serial transactions:

```
// Mathes 'batch' or 'serial' expectations
serial.bwrite_all(&[1, 2]).unwrap();

// Also matches 'batch' or 'serial' expectations
serial.write(1).unwrap();
serial.write(2).unwrap();
```

The comment tries to indicate that the user could write their expectations resembling

```
let expectations = [
    SerialTransaction::read(0x0A),
    SerialTransaction::read(0xFF),
    SerialTransaction::write([1]),
    SerialTransaction::write([2]),
    SerialTransaction::flush(),
];
```